rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,369 @@
+---
+authors: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+
+# RFD 0129 - Avoid Discovery Resource Name Collisions
+
+## Required Approvers
+
+- Engineering: `@r0mant && @smallinsky && @tigrato`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden || @jentfoo`
+
+## What
+
+Auto-Discovery shall name discovered resources such that other resources of
+the same kind are unlikely to have the same name.
+
+In particular, discovered cloud resource names shall include uniquely
+identifying metadata in the name such as region, account ID, or sub-type name.
+
+`tsh` sub-commands shall allow users to use a prefix of the resource name when
+the prefix unambiguously identifies a resource.
+
+Additionally, `tsh` sub-commands shall support using label selectors to
+unambiguously select a single resource.
+
+This RFD does not apply to ssh server instance discovery, since servers are
+already identified within the Teleport cluster by a UUID.
+
+## Why
+
+Multiple discovery agents can discover resources with identical names.
+For example, this happened when customers had databases in different AWS
+regions or accounts with the same name. When a name collision occurs, only one
+of the databases can be accessed by users.
+
+Name collisions can be avoided with the addition of other resource metadata
+in the resource name.
+
+Since discovered resource names will be longer and more tedious to use, we
+should support resource name prefixes and label matching in `tsh` for better UX.
```

Comment:
in `tsh`, web UI and Teleport connect

Reply:
We already support substring and label matching search in web UI and Teleport Connect, I'll add it to the RFD though